lang/LangPrimSource/SC_ComPort.cpp

```diff
@@ -65,16 +66,136 @@ void stopAsioThread() {
+        const double timeReceived = elapsedTime(); // get time now to minimize jitter due to lang load
+
+        ReplyAddress addrObject;
+        addrObject.mProtocol = protocol;
+        addrObject.mAddress = replyAddress;
+        addrObject.mSocket = replySocket;
+        addrObject.mPort = replyPort;
+        addrObject.mReplyFunc = nullptr;
+        addrObject.mReplyData = nullptr;
+
+        ProcessRawMessage(dataSize, std::move(data), addrObject, localPort, timeReceived);
+    }
+};
+
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+namespace Detail {
+
+TCPConnection::TCPConnection(boost::asio::io_service& ioService, int portNum, HandlerType handlerType):
+    mSocket(ioService),
+    mOSCMsgLength(0),
+    mPortNum(portNum) {
+    initHandler(handlerType);
+}
+
+void TCPConnection::initHandler(HandlerType handlerType) {
+    switch (handlerType) {
+    case HandlerType::OSC: {
+        using Handler = MessageHandler<HandlerType::OSC>;
+        mHandleFunc = [this](auto data, auto dataSize) {
+            const int replyPort = 0;
+            const boost::asio::ip::address replyAddress;
+            Handler::handleMessage(kTCP, mSocket.native_handle(), replyAddress, replyPort, std::move(data), dataSize,
+                                   mPortNum);
+        };
+        return;
+    }
+    case HandlerType::Raw: {
+        using Handler = MessageHandler<HandlerType::Raw>;
+        mHandleFunc = [this](auto data, auto dataSize) {
+            const int replyPort = 0;
+            const boost::asio::ip::address replyAddress;
+            Handler::handleMessage(kTCP, mSocket.native_handle(), replyAddress, replyPort, std::move(data), dataSize,
+                                   mPortNum);
+        };
+        return;
+    }
+    }
+}
+
+void TCPConnection::start() {
+    namespace ba = boost::asio;
+    ba::async_read(mSocket, ba::buffer(&mOSCMsgLength, sizeof(mOSCMsgLength)),
+                   [receiver = shared_from_this()](auto error, auto bytesTransferred) {
+                       receiver->handleLengthReceived(error, bytesTransferred);
+                   });
+}
+
+void TCPConnection::handleLengthReceived(const boost::system::error_code& error, size_t bytes_transferred) {
+    if (error) {
+        return;
+    }
+
+    namespace ba = boost::asio;
+    // msglen is in network byte order
+    mOSCMsgLength = sc_ntohl(mOSCMsgLength);
+
+    mData.reset(new char[mOSCMsgLength]);
+
+    ba::async_read(mSocket, ba::buffer(mData.get(), mOSCMsgLength),
+                   [receiver = shared_from_this()](auto error, auto bytesReceived) {
+                       receiver->handleMsgReceived(error, bytesReceived);
+                   });
+}
+
+void TCPConnection::handleMsgReceived(const boost::system::error_code& error, size_t bytes_transferred) {
+    if (error) {
+        mData.reset();
+        return;
+    }
+
+    assert(bytes_transferred == mOSCMsgLength);
```

Comment:
I always like to add a comment to an assert to explain:
- why I might have hit this
- what am I supposed to do if this gets hit

from context here, it's hard for me to know where SC went wrong